src/compiler/scala/tools/nsc/backend/jvm/BCodeBodyBuilder.scala

```diff
@@ -1361,21 +1360,27 @@ abstract class BCodeBodyBuilder extends BCodeSkelBuilder {
     def genLoadTry(tree: Try): BType
 
     def genInvokeDynamicLambda(canLMF: delambdafy.LambdaMetaFactoryCapable) = {
-      import canLMF._
+      import canLMF.{ lambdaTarget => originalTarget, _ }
 
-      val isStaticMethod = lambdaTarget.hasFlag(Flags.STATIC)
+      val lambdaTarget = originalTarget.attachments.get[JustMethodReference].map(_.lambdaTarget).getOrElse(originalTarget)
       def asmType(sym: Symbol) = classBTypeFromSymbol(sym).toASMType
 
       val isInterface = lambdaTarget.owner.isTrait
+      val tag =
+        if (lambdaTarget.isStaticMember) Opcodes.H_INVOKESTATIC
+        else if (lambdaTarget.isPrivate) Opcodes.H_INVOKESPECIAL
+        //else if (lambdaTarget.isClassConstructor) Opcodes.H_NEWINVOKESPECIAL // to invoke Foo::new directly
+        else if (isInterface) Opcodes.H_INVOKEINTERFACE
+        else Opcodes.H_INVOKEVIRTUAL
       val implMethodHandle =
-        new asm.Handle(if (lambdaTarget.hasFlag(Flags.STATIC)) asm.Opcodes.H_INVOKESTATIC else if (isInterface) asm.Opcodes.H_INVOKEINTERFACE else asm.Opcodes.H_INVOKEVIRTUAL,
+        new asm.Handle(
+          tag,
           classBTypeFromSymbol(lambdaTarget.owner).internalName,
           lambdaTarget.name.toString,
           methodBTypeFromSymbol(lambdaTarget).descriptor,
           /* itf = */ isInterface)
-      val receiver = if (isStaticMethod) Nil else lambdaTarget.owner :: Nil
```

Comment:
Was this dead, i.e., was `isStaticMethod` always true?

Reply:
Yeah I think I ran the whole test suite with an assertion and it never threw.